core/embed/trezorhal/stm32u5/trustzone.c

```diff
@@ -96,45 +99,232 @@ static void trustzone_configure_fsmc(void) {
   HAL_FLASHEx_ConfigBBAttributes(&flash_bb);
+  flash_bb.BBAttributesType = FLASH_BB_PRIV;
+  HAL_FLASHEx_ConfigBBAttributes(&flash_bb);
 }
 
-void trustzone_init_boardloader(void) {
+void tz_init_boardloader(void) {
   // Configure ARM SCB/FBU security
-  trustzone_configure_arm();
+  tz_configure_arm();
 
   // Configure SAU security attributes
-  trustzone_configure_sau();
+  tz_configure_sau();
 
   // Enable GTZC (Global Trust-Zone Controller) peripheral clock
   __HAL_RCC_GTZC1_CLK_ENABLE();
   __HAL_RCC_GTZC2_CLK_ENABLE();
 
   // Configure SRAM security attributes
-  trustzone_configure_sram();
+  tz_configure_sram();
+
+  // Configure FLASH security attributes
+  tz_configure_flash();
+
+  // Configure FSMC security attributes
+  tz_configure_fsmc();
+
+  // Make all peripherals secure & privileged
+  HAL_GTZC_TZSC_ConfigPeriphAttributes(
+      GTZC_PERIPH_ALL, GTZC_TZSC_PERIPH_SEC | GTZC_TZSC_PERIPH_PRIV);
+
+  // Clear all illegal access flags in GTZC TZIC
+  HAL_GTZC_TZIC_ClearFlag(GTZC_PERIPH_ALL);
+
+  // Enable all illegal access interrupts in GTZC TZIC
+  HAL_GTZC_TZIC_EnableIT(GTZC_PERIPH_ALL);
+
+  // Enable GTZC secure interrupt
+  NVIC_SetPriority(GTZC_IRQn, IRQ_PRI_HIGHEST);
+  NVIC_EnableIRQ(GTZC_IRQn);
+}
+
+static void set_bit_array(volatile uint32_t* regs, uint32_t bit_offset,
+                          uint32_t bit_count, bool value) {
+  regs += bit_offset / 32;
+  bit_offset %= 32;
+
+  if (bit_offset != 0) {
+    uint32_t bc = MIN(32 - bit_offset, bit_count);
+    uint32_t mask = (1 << bc) - 1;
+
+    mask <<= bit_offset;
+
+    if (value) {
+      *regs |= mask;
+    } else {
+      *regs &= ~mask;
+    }
+
+    regs++;
+    bit_count -= bc;
+  }
+
+  while (bit_count > 32) {
+    *regs = value ? 0xFFFFFFFF : 0;
+    regs++;
+    bit_count -= 32;
+  }
+
+  if (bit_count > 0) {
+    uint32_t mask = (1 << bit_count) - 1;
+    if (value) {
+      *regs |= mask;
+    } else {
+      *regs &= ~mask;
+    }
+  }
+}
+
+typedef struct {
+  // Start address of the region
+  uint32_t start;
+  // End address of the region + 1
+  size_t end;
+  // MPCBB register base
+  volatile GTZC_MPCBB_TypeDef* regs;
+} sram_region_t;
+
+// SRAM regions must be in order of ascending start address
+// and must not overlap
+sram_region_t g_sram_regions[] = {
+    {SRAM1_BASE, SRAM1_BASE + SRAM1_SIZE, GTZC_MPCBB1},
+    {SRAM2_BASE, SRAM2_BASE + SRAM2_SIZE, GTZC_MPCBB2},
+    {SRAM3_BASE, SRAM3_BASE + SRAM3_SIZE, GTZC_MPCBB3},
+#if defined STM32U5A9xx | defined STM32U5G9xx
+    {SRAM5_BASE, SRAM5_BASE + SRAM5_SIZE, GTZC_MPCBB5},
+#endif
+#if defined STM32U5G9xx
+    {SRAM6_BASE, SRAM6_BASE + SRAM6_SIZE, GTZC_MPCBB6},
+#endif
+    {SRAM4_BASE, SRAM4_BASE + SRAM4_SIZE, GTZC_MPCBB4},
+};
+
+void tz_set_sram_unpriv(uint32_t start, uint32_t size, bool unpriv) {
+  const size_t block_size = GTZC_MPCBB_BLOCK_SIZE;
+
+  // Calculate aligned start & end
+  uint32_t end = ((start + size + (block_size - 1)) & ~(block_size - 1));
```

Comment:
I finally modified this so that `tz_sram_set_unpriv()` and `tz_flash_set_unpriv()` no longer extend the address range, and it also `ensures` that the address and size are properly aligned.

The address and size are manually extended in one case (secure_aes unprivileged flash memory) that requires it.

see 9d6c125590a6bba71bb56ff4e3a7ef3cd6ccd8c8 and b35162ac35da5703309fdbc94c6c8df6a4b62cdf